chore(genesys): Clean Up Genesys

jeesooxkim opened the pull request:
Cleaning up Genesys. This PR reorganizes the constants into a similar structure to Twilio Flex, and adds clearer guidance (in the Readme and the comments) on how to add queue-related content.

ugrani commented on src/serviceDesks/genesys/purecloudSetup.ts:
```diff
@@ -14,7 +14,7 @@
   ORGANIZATION_ID = '<your-organization-id>', // https://help.mypurecloud.com/faq/how-do-i-find-my-organization-id/
-  DEPLOYMENT_ID = '<your-deployment-id>', // https://help.mypurecloud.com/articles/create-a-widget-for-web-chat/ ID of widget == DEPLOYMENT_ID
-  QUEUE_TARGET = '<your-queue-name>', // name of queue you want chat requests to go into
+  DEPLOYMENT_ID = '<your-deployment-key>', // https://help.mypurecloud.com/articles/create-a-widget-for-web-chat/ Deployment Key == DEPLOYMENT_ID
```
these things should go to the middleware; the `src/serviceDesks/genesys` is a client side code and should not know anything about server side set up/code

jeesooxkim replied:
These variables don't actually interact with the server code. These are used in the serviceDesk code only as part of messaging, which doesn't require the middleware server to be run.

VikramVuppla commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
+    1. From the `src/middleware/genesys` directory, run `npm install`.
 
-1. In the `example-server` subdirectory, rename or copy `.env-sample` to `.env`.
+    1. From the `src/middleware/genesys` directory, run `npm start`. This starts a server on port 3000 of your local machine.
```
minor comment - change `This starts a server on port 3000 of your local machine` to `This starts a server on port 3000 on your local machine`

VikramVuppla commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
+        **Note:** These flags are independent, so you do not have to enable both simultaneously. For example, if you set `WIDGET_REQUIRES_AUTHENTICATION = false` and `AUTHENTICATED_CALLS_ENABLED = true`, authentication is disabled in the web chat widget, but the server still uses an OAuth token to authenticate when accessing agent availability status.
 
-1. In the Genesys Cloud user interface, open the settings for your web chat widget and enable **Require Authentication**.
+1. Update [`constants.js`](../../middleware/genesys/src/config/constants.js) to populate it with your information.
```
the information should be updated in `.env-sample.js`

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
+1. In Genesys Cloud, [create a widget](https://help.mypurecloud.com/articles/create-a-widget-for-web-chat/) to enable users to chat with agents. Make a note of the generated `Deployment Key`. Follow the instructions for either a `1.1` or `1.0` widget.
 
-1. Update [`purecloudSetup.ts`](../../serviceDesks/genesys/purecloudSetup.ts) to populate it with your information.
+1. Make sure there is an active `Queue` by selecting your profile bubble on the left sidebar and clicking on `Activate Queues`. Check that there is at least one queue under `Active`. If not, either activate an `Available` queue on the right, or [create a queue](https://help.mypurecloud.com/articles/create-queues/). In any case, take note of the name of the `Active` queue you would like to use.
```
```suggestion
1. Make sure there is an active queue by selecting your profile bubble on the left sidebar and clicking on **Activate Queues**. Check that there is at least one queue under **Active**. If not, either activate an available queue on the right, or [create a queue](https://help.mypurecloud.com/articles/create-queues/). In any case, take note of the name of the active queue you would like to use.
```

VikramVuppla commented on src/serviceDesks/genesys/genesysServiceDesk.ts:
```diff
@@ -31,22 +31,27 @@ import {
+ * by AUTH_SERVER_BASE_URL in '../../middleware/genesys/src/config/constants.ts' and use it when creating the chat.
  */
-const WIDGET_REQUIRES_AUTHENTICATION = false;
+const WIDGET_REQUIRES_AUTHENTICATION = true;
```
`WIDGET_REQUIRES_AUTHENTICATION` keep this to false, user can change it according to his needs

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
-1. Update [`purecloudSetup.ts`](../../serviceDesks/genesys/purecloudSetup.ts) to populate it with your information.
+1. Make sure there is an active `Queue` by selecting your profile bubble on the left sidebar and clicking on `Activate Queues`. Check that there is at least one queue under `Active`. If not, either activate an `Available` queue on the right, or [create a queue](https://help.mypurecloud.com/articles/create-queues/). In any case, take note of the name of the `Active` queue you would like to use.
+    - **Creating a queue:**
+      - Name must not include any spaces. This queue name should be unique, and ideally describe the purpose/grouping of the queue. Make sure you click Save.
```
```suggestion
      - The queue name must not include any spaces. This name should be unique, and ideally should describe the purpose or grouping of the queue. Make sure you click **Save**.
```

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
+1. Make sure there is an active `Queue` by selecting your profile bubble on the left sidebar and clicking on `Activate Queues`. Check that there is at least one queue under `Active`. If not, either activate an `Available` queue on the right, or [create a queue](https://help.mypurecloud.com/articles/create-queues/). In any case, take note of the name of the `Active` queue you would like to use.
+    - **Creating a queue:**
+      - Name must not include any spaces. This queue name should be unique, and ideally describe the purpose/grouping of the queue. Make sure you click Save.
+      - Click on the queue that you just created and add yourself through the `Members` tab (you do not have to click save).
```
```suggestion
      - Click on the queue that you just created and add yourself through the Members tab (you do not have to click **Save**).
```

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
 
-  - `ORGANIZATION_ID`: Your Genesys Cloud organization ID. You can find this ID in the [Genesys Cloud settings](https://help.mypurecloud.com/faq/how-do-i-find-my-organization-id/).
-  - `DEPLOYMENT_ID`: The ID of the widget you created in the previous step.
+1. Set up OAuth Client server:
```
```suggestion
1. Set up an OAuth Client server:
```

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
+    - **Creating a queue:**
+      - Name must not include any spaces. This queue name should be unique, and ideally describe the purpose/grouping of the queue. Make sure you click Save.
+      - Click on the queue that you just created and add yourself through the `Members` tab (you do not have to click save).
+      - Any newly created Queues will be `Active` by default.
```
```suggestion
      - Any newly created Queues will be active by default.
```

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
+    1. From the `src/middleware/genesys` directory, run `npm start`. This starts a server on port 3000 of your local machine.
 
-1. In the `.env` file, update the values of the `GENESYS_CLIENT_ID` and `GENESYS_CLIENT_SECRET` to the credentials from your Genesys OAuth client.
+    1. The server needs to be accessed from the browser of your end users. If you do not have your own hosted environment and you wish to expose your local development for testing, consider using a service such as [ngrok](https://ngrok.com/) to create a public URL:
```
```suggestion
    1. Your end users must be able to access the server from their browsers. If you do not have your own hosted environment, and you wish to expose your local development for testing, consider using a service such as [ngrok](https://ngrok.com/) to create a public URL:
```

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
 
-1. In the **Authentication URL** field, specify `https://<server-url>/jwt`, where `<server-url>` is the public URL for your server. Click **Save**.
+    - `ORGANIZATION_ID`: Your Genesys Cloud organization ID. This ID is visible in your widget configuration page in the `Generated Script Tag` block's `org-guid` field.  You can also find this ID in the [Genesys Cloud settings](https://help.mypurecloud.com/faq/how-do-i-find-my-organization-id/).
+    - `DEPLOYMENT_ID`: The `Deployment Key` of the widget you created in the previous steps.
```
```suggestion
    - `DEPLOYMENT_ID`: The deployment key of the widget you created in the previous steps.
```

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
-1. In the **Authentication URL** field, specify `https://<server-url>/jwt`, where `<server-url>` is the public URL for your server. Click **Save**.
+    - `ORGANIZATION_ID`: Your Genesys Cloud organization ID. This ID is visible in your widget configuration page in the `Generated Script Tag` block's `org-guid` field.  You can also find this ID in the [Genesys Cloud settings](https://help.mypurecloud.com/faq/how-do-i-find-my-organization-id/).
+    - `DEPLOYMENT_ID`: The `Deployment Key` of the widget you created in the previous steps.
+    - `QUEUE_TARGET`: The name of your active Queue, noted from the previous steps.
```
```suggestion
    - `QUEUE_TARGET`: The name of your active queue, noted from the previous steps.
```

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
+    - `ORGANIZATION_ID`: Your Genesys Cloud organization ID. This ID is visible in your widget configuration page in the `Generated Script Tag` block's `org-guid` field.  You can also find this ID in the [Genesys Cloud settings](https://help.mypurecloud.com/faq/how-do-i-find-my-organization-id/).
+    - `DEPLOYMENT_ID`: The `Deployment Key` of the widget you created in the previous steps.
+    - `QUEUE_TARGET`: The name of your active Queue, noted from the previous steps.
+    - `AUTH_SERVER_BASE_URL`: The public URL for your server - if you used `ngrok` above for your public URL, this should look like `https://<some-hash>.ngrok.io`. Make sure you specify the `https://` URL.
```
```suggestion
    - `AUTH_SERVER_BASE_URL`: The public URL of your server. If you used `ngrok` to create your public URL, this looks like `https://<some-hash>.ngrok.io`. Make sure you specify the secure `https://` URL.
```

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,56 @@ This reference implementation supports the core features of a Genesys Cloud inte
+1. Go to the project root directory and edit the `.env` file. Update the `SERVICE_DESK_CLASS` variable to `GenesysServiceDesk`.
 
-1. From the project root directory, run `npm run dev` again, and test. The server should now authenticate the chat and fetch agent availability.
+1. From the project root directory, run `npm run dev`. To receive requests, make sure you are `On Queue` (top right button) in your Genesys account. If you've linked everything to your Genesys Cloud account correctly, you should be able to connect to an agent in Genesys Cloud, fully functional with authenticated chat and agent availability.
```
```suggestion
1. From the project root directory, run `npm run dev`. To receive requests, make sure you have clicked the **On Queue** button in the Genesys UI. If you've linked everything to your Genesys Cloud account correctly, you should be able to connect to an agent in Genesys Cloud, fully functional with authenticated chat and agent availability.
```

rrberry commented:
Made some suggestions for the README update to comply with various IBM style and editorial guidelines.

VikramVuppla reviewed (commented):
currently the genesys code is written in javascript, it would be super helpful and consistent across service desks, if its modified to Typescript

VikramVuppla commented:
with the change in folder structure, the arch diagram needs to be updated as well https://github.com/watson-developer-cloud/assistant-web-chat-service-desk-starter/blob/main/src/middleware/genesys/arch-overview.png

VikramVuppla commented on src/serviceDesks/genesys/genesysServiceDesk.ts:
```diff
@@ -124,33 +123,49 @@ class GenesysServiceDesk implements ServiceDesk {
-      deploymentId: setup.DEPLOYMENT_ID,
+    let env;
+    try {
+      const setupEnv = await fetch(`http://localhost:3000/setup`);
```
extract as an env var `SERVER_BASE_URL`

VikramVuppla reviewed (commented):
verify use cases for `WIDGET_REQUIRES_AUTHENTICATION` and `AUTHENTICATED_CALLS_ENABLED` are working correctly

fix the failing middleware server start

jeesooxkim commented:
@VikramVuppla Verified functionality for toggles, and actually renamed them. I removed `AUTHENTICATED_CALLS_ENABLED`  since that was an artifact from developing the OAuth required API calls. I renamed `WIDGET_REQUIRES_AUTHENTICATION` to `REQUIRES_USER_AUTHENTICATION`, which toggles on/off the JWT authenticated chat. So in short, now there is only a single toggle, for JWT authenticated chat. Agent availability is now always on. I've updated the README as well with clearer instructions.

I can't duplicate your module not found issue with the middleware server, but it may be a node/npm version issue. My node version is `14.4.0`, and npm is `6.14.4`.

VikramVuppla reviewed (dismissed):
LGTM

ugrani reviewed (dismissed):
LGTM

VikramVuppla commented:
@rrberry could you do one final review on the docs/[Readme](https://github.com/jeesooxkim/assistant-web-chat-service-desk-starter/blob/main/src/middleware/genesys/README.md)

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,73 @@ This reference implementation supports the core features of a Genesys Cloud inte
 1. If you haven't done so already, follow the setup steps in the root-level [README](../../../README.md) to make sure you can run an instance of [ExampleServiceDesk](../../serviceDesks/exampleServiceDesk.ts).
 
-1. In Genesys Cloud, [create a widget](https://help.mypurecloud.com/articles/create-a-widget-for-web-chat/) to enable users to chat with agents. Make a note of the widget ID.
+1. In Genesys Cloud, [create a widget](https://help.mypurecloud.com/articles/create-a-widget-for-web-chat/) to enable users to chat with agents. Make a note of the generated `Deployment Key`. Follow the instructions for either a `1.1` or `1.0` widget.
```
```suggestion
1. In Genesys Cloud, [create a widget](https://help.mypurecloud.com/articles/create-a-widget-for-web-chat/) to enable users to chat with agents. Make a note of the generated deployment key. Follow the instructions for either a `1.1` or `1.0` widget.
```

rrberry commented on src/middleware/genesys/README.md:
```diff
@@ -17,53 +17,73 @@ This reference implementation supports the core features of a Genesys Cloud inte
+        - `QUEUE_TARGET`: The name of your active queue, noted from the previous steps.
 
-To configure advanced functionality such as authenticated chat and agent availability status, follow these steps:
+    1. (Optional) Set up advanced features (Agent Availability, Authenticated Chat).
```
```suggestion
    1. (Optional) Set up advanced features (agent availability and authenticated chat).
```

rrberry commented:
Spotted a couple more minor style issues. Everything else looks fine, so I will approve as soon as those last ones are dealt with.

jeesooxkim commented:
@rrberry Added your suggested edits in the README.

VikramVuppla reviewed (approved):
LGTM
